src/node_crypto.cc

```diff
@@ -2579,6 +2580,839 @@ static bool IsSupportedAuthenticatedMode(const EVP_CIPHER_CTX* ctx) {
+  str->WriteUtf8(env->isolate(), data, alloc_size, nullptr, opts);
+  return Allocated(data, size);
+}
+
+ByteSource ByteSource::FromBuffer(Local<Value> buffer, bool ntc) {
+  size_t size = Buffer::Length(buffer);
+  if (ntc) {
+    char* data = MallocOpenSSL<char>(size + 1);
+    memcpy(data, Buffer::Data(buffer), size);
+    data[size] = 0;
+    return Allocated(data, size);
+  }
+  return Foreign(Buffer::Data(buffer), size);
+}
+
+ByteSource ByteSource::NullTerminatedCopy(Environment* env,
+                                          Local<Value> value) {
+  return Buffer::HasInstance(value) ? FromBuffer(value, true)
+                                    : FromString(env, value.As<String>(), true);
+}
+
+ByteSource ByteSource::FromSymmetricKeyObject(Local<Value> handle) {
+  CHECK(handle->IsObject());
+  KeyObject* key = Unwrap<KeyObject>(handle.As<Object>());
+  CHECK(key);
+  // Note that this is only safe as long as the key object cannot be accessed
+  // by other threads.
+  return Foreign(key->GetSymmetricKey(), key->GetSymmetricKeySize());
+}
+
+ByteSource::ByteSource(const char* data, char* allocated_data, size_t size)
+      : data_(data),
+        allocated_data_(allocated_data),
+        size_(size) {}
+
+ByteSource ByteSource::Allocated(char* data, size_t size) {
+  return ByteSource(data, data, size);
+}
+
+ByteSource ByteSource::Foreign(const char* data, size_t size) {
+  return ByteSource(data, nullptr, size);
+}
+
+enum KeyEncodingContext {
+  kKeyContextInput,
+  kKeyContextExport,
+  kKeyContextGenerate
+};
+
+static void GetKeyFormatAndTypeFromJS(
+    AsymmetricKeyEncodingConfig* config,
+    const FunctionCallbackInfo<Value>& args,
+    unsigned int* offset,
+    KeyEncodingContext context) {
+  // During key pair generation, it is possible not to specify a key encoding,
+  // which will lead to a key object being returned.
+  if (args[*offset]->IsUndefined()) {
+    CHECK_EQ(context, kKeyContextGenerate);
+    CHECK(args[*offset + 1]->IsUndefined());
+    config->output_key_object_ = true;
+  } else {
+    config->output_key_object_ = false;
+
+    CHECK(args[*offset]->IsInt32());
+    config->format_ = static_cast<PKFormatType>(
+        args[*offset].As<Int32>()->Value());
+
+    if (args[*offset + 1]->IsInt32()) {
+      config->type_ = Just<PKEncodingType>(static_cast<PKEncodingType>(
+          args[*offset + 1].As<Int32>()->Value()));
+    } else {
+      CHECK(context == kKeyContextInput && config->format_ == kKeyFormatPEM);
+      CHECK(args[*offset + 1]->IsNullOrUndefined());
+      config->type_ = Nothing<PKEncodingType>();
+    }
+  }
+
+  *offset += 2;
+}
+
+static PublicKeyEncodingConfig GetPublicKeyEncodingFromJS(
+    const FunctionCallbackInfo<Value>& args,
+    unsigned int* offset,
+    KeyEncodingContext context) {
+  PublicKeyEncodingConfig result;
+  GetKeyFormatAndTypeFromJS(&result, args, offset, context);
+  return result;
+}
+
+static ManagedEVPPKey GetPublicKeyFromJS(
```

Comment:
I'm a big fan of treating acronyms as words for the purposes of CamelCase: `ManagedEvppKey GetPublicKeyFromJs()`... PKCS1OAEPPSSAES ... argh. Ignore if that's not the node c++ way, or you don't agree.